ACM Digital Library.js

```diff
@@ -1,203 +1,200 @@
-			for (var i=0; i<item.tags.length; i++) {
-				item.tags[i] = item.tags[i].replace("</kwd>", "");
-			}
-			
-			//full issues of journals/magazines don't have a title
-			if (!item.title && text.includes("issue_date")) {
-				var m = text.match(/issue_date\s*=\s*{(.*)},?/);
-				item.itemType = "book";
-				item.title = item.publicationTitle;
-				if (m) {
-					item.title = item.title + ", " + m[1];
-				}
+	let extractedContext = doc.querySelector('meta[name=pbContext]').content;
+	let doiRegex = /article:article:doi\\:([^;]+)/;
+	let doi = extractedContext.match(doiRegex)[1].toLowerCase();
+	let lookupEndpoint = 'https://dl.acm.org/action/exportCiteProcCitation?targetFile=custom-bibtex&format=bibTex&dois=';
+	ZU.doGet(lookupEndpoint + doi, function (text) {
+		let json = JSON.parse(text);
+		let cslItem = json.items[0][doi];
+		cslItem.type = cslItem.type.toLowerCase().replace('_', '-');
+		
+		// For theses, the advisor is indicated as an editor in CSL which
+		// ZU.itemFromCSLJSON incorrectly extracts as an author.
+		if (cslItem.type == 'thesis') {
+			delete cslItem.editor;
+		}
+		
+		let item = new Zotero.Item();
+		ZU.itemFromCSLJSON(item, cslItem);
+		
+		let abstractElements = doc.querySelectorAll('div.article__abstract p, div.abstractSection p');
+		let abstract = Array.from(abstractElements).map(function (element) {
+			return element.textContent;
+		}).join(' ');
+		if (abstract.length && abstract.toLowerCase() != 'no abstract available.') {
+			item.abstractNote = ZU.trimInternal(abstract);
+		}
+		
+		let pdfElement = doc.querySelector('a[title=PDF]');
+		if (pdfElement) {
+			item.attachments.push({
+				url: pdfElement.href,
+				title: 'Full Text PDF',
+				mimeType: 'application/pdf'
+			});
+		}
+		
+		if (item.itemType == 'journalArticle') {
+			// Publication name in the CSL is shortened; scrape from page to get full title.
+			let expandedTitle = doc.querySelector('span.epub-section__title');
+			if (expandedTitle) {
+				item.publicationTitle = expandedTitle.textContent;
 			}
-			
-			//The abstract from above or we try to make an individual request
-			//e.g. for multiples
-			if (!item.abstractNote) {
-				if (abs && abs.trim()) {
-					item.abstractNote = abs;
-					item.complete();
-				} else {
-					ZU.doGet("https://dl.acm.org/tab_abstract.cfm?id="+itemID, function(abstract) {
-						item.abstractNote = ZU.unescapeHTML(abstract);
-						if (item.abstractNote.trim() == "An abstract is not available.") delete item.abstractNote;
-						item.complete();
-					});
-				}
-			} else {
-				item.complete();
+		}
+		
+		if (item.itemType == 'bookSection' && item.creators.length == 0) {
+			// A chapter of a book is extracted as a bookSection, but the
+			// authors are not included in the CSL so we must scrape them.
+			// e.g. https://dl.acm.org/doi/abs/10.5555/3336323.C5474411
+			let authorElements = doc.querySelectorAll('div.citation span.loa__author-name');
```

Comment:
We could do this always when `item.creators` array is empty but `authorElements` not, i.e. there is no reason to restrict this to bookSections only.